components/base/kinematicbase/differentialDrive.go

```diff
@@ -320,3 +309,46 @@ func (ddk *differentialDriveKinematics) newValidRegionCapsule(starting, desired
 
 	return capsule, nil
 }
+
+//nolint: dupl
+func (ddk *differentialDriveKinematics) ErrorState(
+	ctx context.Context,
+	plan [][]referenceframe.Input,
+	currentNode int,
+) (spatialmath.Pose, error) {
+	if currentNode < 0 || currentNode >= len(plan) {
+		return nil, fmt.Errorf("cannot get ErrorState for node %d, must be >= 0 and less than plan length %d", currentNode, len(plan))
+	}
+
+	// Get pose-in-frame of the base via its localizer. The offset between the localizer and its base should already be accounted for.
+	actualPIF, err := ddk.CurrentPosition(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var nominalPose spatialmath.Pose
+
+	// Determine the nominal pose, that is, the pose where the robot ought be if it had followed the plan perfectly up until this point.
+	// This is done differently depending on what sort of frame we are working with.
+	if len(plan) < 2 {
+		return nil, errors.New("diff drive motion plan must have at least two waypoints")
+	}
+	nominalPose, err = ddk.planningFrame.Transform(plan[currentNode])
+	if err != nil {
+		return nil, err
+	}
+	if currentNode > 0 {
+		pastPose, err := ddk.planningFrame.Transform(plan[currentNode-1])
+		if err != nil {
+			return nil, err
+		}
+		// diff drive bases don't have a notion of "distance along the trajectory between waypoints", so instead we compare to the
+		// nearest point on the straight line path.
+		nominalPoint := spatialmath.ClosestPointSegmentPoint(pastPose.Point(), nominalPose.Point(), actualPIF.Pose().Point())
+		pointDiff := nominalPose.Point().Sub(pastPose.Point())
+		desiredHeading := math.Atan2(pointDiff.Y, pointDiff.X)
```

Comment:
I originally solved this with frame systems and I think thats a little less error prone than doing angle math in this function

Reply:
This is not the same query as where the frame system was being used; that was equivalent to a `PoseBetween` call.

This is figuring out the desired heading based on the last and current waypoints, in the same way we do already above on line 129 and 287.

Reply:
Ok sounds good with me then